// @progress/kendo-react-treelist/dist/npm/cells/FilterCells/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.localizeFilter = exports.booleanFilterValues = exports.booleanOperator = exports.dateFilterOperators = exports.numericFilterOperators = exports.textFilterOperators = void 0;
var loc = require("../../messages");
/**
 * @hidden
 */
exports.textFilterOperators = [
    { text: loc.filterContainsOperator, operator: 'contains' },
    { text: loc.filterNotContainsOperator, operator: 'doesnotcontain' },
    { text: loc.filterEqOperator, operator: 'eq' },
    { text: loc.filterNotEqOperator, operator: 'neq' },
    { text: loc.filterStartsWithOperator, operator: 'startswith' },
    { text: loc.filterEndsWithOperator, operator: 'endswith' },
    { text: loc.filterIsNullOperator, operator: 'isnull' },
    { text: loc.filterIsNotNullOperator, operator: 'isnotnull' },
    { text: loc.filterIsEmptyOperator, operator: 'isempty' },
    { text: loc.filterIsNotEmptyOperator, operator: 'isnotempty' }
];
/**
 * @hidden
 */
exports.numericFilterOperators = [
    { text: loc.filterEqOperator, operator: 'eq' },
    { text: loc.filterNotEqOperator, operator: 'neq' },
    { text: loc.filterGteOperator, operator: 'gte' },
    { text: loc.filterGtOperator, operator: 'gt' },
    { text: loc.filterLteOperator, operator: 'lte' },
    { text: loc.filterLtOperator, operator: 'lt' },
    { text: loc.filterIsNullOperator, operator: 'isnull' },
    { text: loc.filterIsNotNullOperator, operator: 'isnotnull' }
];
/**
 * @hidden
 */
exports.dateFilterOperators = [
    { text: loc.filterEqOperator, operator: 'eq' },
    { text: loc.filterNotEqOperator, operator: 'neq' },
    { text: loc.filterAfterOrEqualOperator, operator: 'gte' },
    { text: loc.filterAfterOperator, operator: 'gt' },
    { text: loc.filterBeforeOperator, operator: 'lt' },
    { text: loc.filterBeforeOrEqualOperator, operator: 'lte' },
    { text: loc.filterIsNullOperator, operator: 'isnull' },
    { text: loc.filterIsNotNullOperator, operator: 'isnotnull' }
];
/**
 * @hidden
 */
exports.booleanOperator = 'eq';
/**
 * @hidden
 */
exports.booleanFilterValues = [
    { text: loc.filterBooleanAll, operator: '' },
    { text: loc.filterIsTrue, operator: true },
    { text: loc.filterIsFalse, operator: false }
];
/**
 * @hidden
 */
function localizeFilter(localization, operators) {
    return {
        clearButtonTitle: localization.toLanguageString(loc.filterClearButton, loc.messages[loc.filterClearButton]),
        operators: operators.map(function (o) { return ({
            text: localization.toLanguageString(o.text, loc.messages[o.text]),
            operator: o.operator
        }); })
    };
}
exports.localizeFilter = localizeFilter;
