// @progress/kendo-react-listbox/dist/npm/messages/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.messages = exports.remove = exports.transferAllFrom = exports.transferAllTo = exports.transferFrom = exports.transferTo = exports.moveDown = exports.moveUp = void 0;
/**
 * @hidden
 */
exports.moveUp = 'listbox.moveUp';
/**
 * @hidden
 */
exports.moveDown = 'listbox.moveDown';
/**
 * @hidden
 */
exports.transferTo = 'listbox.transferTo';
/**
 * @hidden
 */
exports.transferFrom = 'listbox.transferFrom';
/**
 * @hidden
 */
exports.transferAllTo = 'listbox.transferAllTo';
/**
 * @hidden
 */
exports.transferAllFrom = 'listbox.transferAllFrom';
/**
 * @hidden
 */
exports.remove = 'listbox.remove';
/**
 * @hidden
 */
exports.messages = (_a = {},
    _a[exports.moveUp] = 'Move Up',
    _a[exports.moveDown] = 'Move Down',
    _a[exports.transferTo] = 'Transfer To',
    _a[exports.transferFrom] = 'Transfer From',
    _a[exports.transferAllTo] = 'Transfer All To',
    _a[exports.transferAllFrom] = 'Transfer All From',
    _a[exports.remove] = 'Delete',
    _a);
