// @progress/kendo-react-layout/dist/npm/menu/components/MenuItemArrow.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MenuItemArrow = exports.leftArrowName = exports.rightArrowName = exports.downArrowName = void 0;
var React = require("react");
var PropTypes = require("prop-types");
var kendo_react_common_1 = require("@progress/kendo-react-common");
var kendo_svg_icons_1 = require("@progress/kendo-svg-icons");
var misc_1 = require("../utils/misc");
/**
 * @hidden
 */
exports.downArrowName = 'caret-alt-down';
/**
 * @hidden
 */
exports.rightArrowName = 'caret-alt-right';
/**
 * @hidden
 */
exports.leftArrowName = 'caret-alt-left';
var MenuItemArrow = /** @class */ (function (_super) {
    __extends(MenuItemArrow, _super);
    function MenuItemArrow() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @hidden
     */
    MenuItemArrow.prototype.render = function () {
        return React.createElement(kendo_react_common_1.IconWrap, __assign({ "aria-hidden": true, className: 'k-menu-expand-arrow' }, this.getIcon()));
    };
    MenuItemArrow.prototype.getIcon = function () {
        var childrenPosition = (0, misc_1.getChildrenPosition)(this.props.itemId, this.props.verticalMenu === true, this.props.dir === 'rtl');
        switch (childrenPosition) {
            case 'downward':
                return { name: exports.downArrowName, icon: kendo_svg_icons_1.caretAltDownIcon };
            case 'rightward':
                return { name: exports.rightArrowName, icon: kendo_svg_icons_1.caretAltRightIcon };
            case 'leftward':
                return { name: exports.leftArrowName, icon: kendo_svg_icons_1.caretAltLeftIcon };
            default:
                return {};
        }
    };
    /**
     * @hidden
     */
    MenuItemArrow.propTypes = {
        itemId: PropTypes.string,
        dir: PropTypes.string,
        verticalMenu: PropTypes.bool
    };
    return MenuItemArrow;
}(React.Component));
exports.MenuItemArrow = MenuItemArrow;
