// @progress/kendo-react-gauges/dist/npm/ArcGauge.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArcGauge = void 0;
var React = require("react");
var BaseGauge_1 = require("./BaseGauge");
var kendo_charts_1 = require("@progress/kendo-charts");
var ArcCenter_1 = require("./ArcCenter");
var kendo_drawing_1 = require("@progress/kendo-drawing");
var kendo_react_common_1 = require("@progress/kendo-react-common");
var ArcGauge = /** @class */ (function (_super) {
    __extends(ArcGauge, _super);
    function ArcGauge() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this._baseGauge = null;
        _this.gaugeType = kendo_charts_1.ArcGauge;
        _this.getTarget = function () {
            return _this;
        };
        _this.deriveOptionsFromParent = function (options) {
            var _a = _this.props, value = _a.value, color = _a.color, colors = _a.colors, opacity = _a.opacity, scale = _a.scale;
            return Object.assign({}, options, {
                value: value,
                color: color,
                colors: colors,
                opacity: opacity,
                scale: scale
            });
        };
        return _this;
    }
    Object.defineProperty(ArcGauge.prototype, "gaugeInstance", {
        /**
         * @hidden
         */
        get: function () {
            if (this._baseGauge) {
                return this._baseGauge.gaugeInstance;
            }
            return null;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ArcGauge.prototype, "surface", {
        /**
         * The Drawing `Surface` of the Gauge.
         */
        get: function () {
            if (this._baseGauge) {
                return this._baseGauge.surface;
            }
            return null;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ArcGauge.prototype, "element", {
        /**
         * The DOM element of the Gauge.
         */
        get: function () {
            if (this._baseGauge) {
                return this._baseGauge.element;
            }
            return null;
        },
        enumerable: false,
        configurable: true
    });
    /**
     * @hidden
     */
    ArcGauge.prototype.render = function () {
        var _this = this;
        var _a = this.props, children = _a.children, arcCenterRender = _a.arcCenterRender, centerRender = _a.centerRender, className = _a.className, baseGaugeProps = __rest(_a, ["children", "arcCenterRender", "centerRender", "className"]);
        return (React.createElement(BaseGauge_1.BaseGauge, __assign({}, baseGaugeProps, { deriveOptionsFromParent: this.deriveOptionsFromParent, ref: function (baseGauge) { return _this._baseGauge = baseGauge; }, gaugeConstructor: this.gaugeType, getTarget: this.getTarget, className: (0, kendo_react_common_1.classNames)('k-gauge k-arcgauge', className) }),
            children,
            React.createElement(ArcCenter_1.ArcCenter, { render: centerRender || arcCenterRender })));
    };
    /**
     * Exports a Gauge component as a Drawing `Scene`.
     *
     * @param {any} options - The parameters for the export operation.
     * @returns {Promise<Group>} - A promise that returns the root `Group` of the scene.
     */
    ArcGauge.prototype.exportVisual = function (options) {
        if (this.gaugeInstance !== null && this.element) {
            return (0, kendo_drawing_1.drawDOM)(this.element, options);
        }
        return Promise.resolve(new kendo_drawing_1.Group());
    };
    return ArcGauge;
}(React.Component));
exports.ArcGauge = ArcGauge;
