// @progress/kendo-react-gantt/dist/npm/context/GanttContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GanttContext = exports.useGanttTaskEventsContext = exports.GanttTaskEventsContext = exports.useGanttEventsContext = exports.GanttEventsContext = exports.useGanttToolbarHeightContext = exports.GanttToolbarHeightContext = exports.useGanttRowHeightContext = exports.GanttRowHeightContext = exports.useGanttActiveViewContext = exports.GanttActiveViewContext = exports.useGanttViewsContext = exports.GanttViewsContext = exports.useGanttDateRangeContext = exports.GanttDateRangeContext = exports.useGanttPropsContext = exports.GanttPropsContext = exports.useGanttDependencyModelFieldsContext = exports.GanttDependencyModelFieldsContext = exports.useGanttTaskModelFieldsContext = exports.GanttTaskModelFieldsContext = exports.useGanttDependencyDataContext = exports.GanttDependencyDataContext = exports.useGanttTaskDataContext = exports.GanttTaskDataContext = void 0;
var React = require("react");
var kendo_date_math_1 = require("@progress/kendo-date-math");
var kendo_react_common_1 = require("@progress/kendo-react-common");
var constants_1 = require("../constants");
/** @hidden */
exports.GanttTaskDataContext = React.createContext([]);
/** @hidden */
var useGanttTaskDataContext = function () { return React.useContext(exports.GanttTaskDataContext); };
exports.useGanttTaskDataContext = useGanttTaskDataContext;
exports.GanttTaskDataContext.displayName = 'KendoReactGanttTaskDataContext';
/** @hidden */
exports.GanttDependencyDataContext = React.createContext([]);
/** @hidden */
var useGanttDependencyDataContext = function () { return React.useContext(exports.GanttDependencyDataContext); };
exports.useGanttDependencyDataContext = useGanttDependencyDataContext;
exports.GanttDependencyDataContext.displayName = 'KendoReactGanttDependencyDataContext';
/** @hidden */
exports.GanttTaskModelFieldsContext = React.createContext(constants_1.DEFAULT_TASK_MODEL_FIELDS);
/** @hidden */
var useGanttTaskModelFieldsContext = function () { return React.useContext(exports.GanttTaskModelFieldsContext); };
exports.useGanttTaskModelFieldsContext = useGanttTaskModelFieldsContext;
exports.GanttTaskModelFieldsContext.displayName = 'KendoReactGanttTaskModelFieldsContext';
/** @hidden */
exports.GanttDependencyModelFieldsContext = React.createContext(constants_1.DEFAULT_DEPENDENCY_MODEL_FIELDS);
/** @hidden */
var useGanttDependencyModelFieldsContext = function () { return React.useContext(exports.GanttDependencyModelFieldsContext); };
exports.useGanttDependencyModelFieldsContext = useGanttDependencyModelFieldsContext;
exports.GanttDependencyModelFieldsContext.displayName = 'KendoReactGanttDependencyModelFieldsContext';
/** @hidden */
exports.GanttPropsContext = React.createContext({});
/** @hidden */
var useGanttPropsContext = function () { return React.useContext(exports.GanttPropsContext); };
exports.useGanttPropsContext = useGanttPropsContext;
exports.GanttPropsContext.displayName = 'KendoReactGanttPropsContext';
/** @hidden */
exports.GanttDateRangeContext = React.createContext({
    start: new Date(),
    end: new Date(),
    zonedStart: kendo_date_math_1.ZonedDate.fromLocalDate(new Date()),
    zonedEnd: kendo_date_math_1.ZonedDate.fromLocalDate(new Date())
});
/** @hidden */
var useGanttDateRangeContext = function () { return React.useContext(exports.GanttDateRangeContext); };
exports.useGanttDateRangeContext = useGanttDateRangeContext;
exports.GanttDateRangeContext.displayName = 'KendoReactGanttDateRangeContext';
/** @hidden */
exports.GanttViewsContext = React.createContext([]);
/** @hidden */
var useGanttViewsContext = function () { return React.useContext(exports.GanttViewsContext); };
exports.useGanttViewsContext = useGanttViewsContext;
exports.GanttViewsContext.displayName = 'KendoReactGanttViewsContext';
/** @hidden */
exports.GanttActiveViewContext = React.createContext(['week', kendo_react_common_1.noop]);
/** @hidden */
var useGanttActiveViewContext = function () { return React.useContext(exports.GanttActiveViewContext); };
exports.useGanttActiveViewContext = useGanttActiveViewContext;
exports.GanttActiveViewContext.displayName = 'KendoReactGanttActiveViewContext';
/** @hidden */
exports.GanttRowHeightContext = React.createContext(0);
/** @hidden */
var useGanttRowHeightContext = function () { return React.useContext(exports.GanttRowHeightContext); };
exports.useGanttRowHeightContext = useGanttRowHeightContext;
exports.GanttRowHeightContext.displayName = 'KendoReactGanttRowHeightContext';
/** @hidden */
exports.GanttToolbarHeightContext = React.createContext(0);
/** @hidden */
var useGanttToolbarHeightContext = function () { return React.useContext(exports.GanttToolbarHeightContext); };
exports.useGanttToolbarHeightContext = useGanttToolbarHeightContext;
exports.GanttToolbarHeightContext.displayName = 'KendoReactGanttToolbarHeightContext';
/** @hidden */
exports.GanttEventsContext = React.createContext({
    onDataStateChange: kendo_react_common_1.noop,
    onSortChange: kendo_react_common_1.noop,
    onFilterChange: kendo_react_common_1.noop,
    onColumnMenuFilterChange: kendo_react_common_1.noop,
    onExpandChange: kendo_react_common_1.noop,
    onRowClick: kendo_react_common_1.noop,
    onRowDoubleClick: kendo_react_common_1.noop,
    onRowContextMenu: kendo_react_common_1.noop,
    onColumnResize: kendo_react_common_1.noop,
    onColumnReorder: kendo_react_common_1.noop,
    onDependencyCreate: kendo_react_common_1.noop,
    onKeyDown: kendo_react_common_1.noop,
    onSelectionChange: kendo_react_common_1.noop,
    onHeaderSelectionChange: kendo_react_common_1.noop
});
/** @hidden */
var useGanttEventsContext = function () { return React.useContext(exports.GanttEventsContext); };
exports.useGanttEventsContext = useGanttEventsContext;
exports.GanttEventsContext.displayName = 'KendoReactGanttEventsContext';
/** @hidden */
exports.GanttTaskEventsContext = React.createContext({
    onTaskClick: kendo_react_common_1.noop,
    onTaskDoubleClick: kendo_react_common_1.noop,
    onTaskContextMenu: kendo_react_common_1.noop
});
/** @hidden */
var useGanttTaskEventsContext = function () { return React.useContext(exports.GanttTaskEventsContext); };
exports.useGanttTaskEventsContext = useGanttTaskEventsContext;
exports.GanttTaskEventsContext.displayName = 'KendoReactGanttTaskEventsContext';
/** @hidden */
var GanttContext = function (_a) {
    var children = _a.children, taskData = _a.taskData, dependencyData = _a.dependencyData, taskModelFields = _a.taskModelFields, dependencyModelFields = _a.dependencyModelFields, props = _a.props, dateRange = _a.dateRange, views = _a.views, activeView = _a.activeView, rowHeight = _a.rowHeight, toolbarHeight = _a.toolbarHeight, events = _a.events, taskEvents = _a.taskEvents;
    return (React.createElement(exports.GanttPropsContext.Provider, { value: props },
        React.createElement(exports.GanttTaskDataContext.Provider, { value: taskData },
            React.createElement(exports.GanttDependencyDataContext.Provider, { value: dependencyData },
                React.createElement(exports.GanttTaskModelFieldsContext.Provider, { value: taskModelFields },
                    React.createElement(exports.GanttDependencyModelFieldsContext.Provider, { value: dependencyModelFields },
                        React.createElement(exports.GanttDateRangeContext.Provider, { value: dateRange },
                            React.createElement(exports.GanttViewsContext.Provider, { value: views },
                                React.createElement(exports.GanttActiveViewContext.Provider, { value: activeView },
                                    React.createElement(exports.GanttRowHeightContext.Provider, { value: rowHeight },
                                        React.createElement(exports.GanttToolbarHeightContext.Provider, { value: toolbarHeight },
                                            React.createElement(exports.GanttEventsContext.Provider, { value: events },
                                                React.createElement(exports.GanttTaskEventsContext.Provider, { value: taskEvents }, children)))))))))))));
};
exports.GanttContext = GanttContext;
exports.GanttContext.displayName = 'KendoReactGanttContext';
