// @progress/kendo-react-gantt/dist/npm/context/GanttViewContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GanttViewContext = exports.useGanttViewDependencyDragContext = exports.GanttViewDependencyDragContext = exports.useGanttViewTimelineWidthContext = exports.GanttViewTimelineWidthContext = exports.useGanttViewTasksContext = exports.GanttViewTasksContext = void 0;
var React = require("react");
var kendo_react_common_1 = require("@progress/kendo-react-common");
/** @hidden */
exports.GanttViewTasksContext = React.createContext([{ current: {} }, kendo_react_common_1.noop]);
/** @hidden */
var useGanttViewTasksContext = function () { return React.useContext(exports.GanttViewTasksContext); };
exports.useGanttViewTasksContext = useGanttViewTasksContext;
exports.GanttViewTasksContext.displayName = 'KendoReactGanttViewContext';
/** @hidden */
exports.GanttViewTimelineWidthContext = React.createContext(0);
/** @hidden */
var useGanttViewTimelineWidthContext = function () { return React.useContext(exports.GanttViewTimelineWidthContext); };
exports.useGanttViewTimelineWidthContext = useGanttViewTimelineWidthContext;
exports.GanttViewTimelineWidthContext.displayName = 'KendoReactGanttViewTimelineWidthContext';
/** @hidden */
exports.GanttViewDependencyDragContext = React.createContext({
    onDependencyHandlePress: kendo_react_common_1.noop,
    onDependencyHandleDrag: kendo_react_common_1.noop,
    onDependencyHandleRelease: kendo_react_common_1.noop,
    isEnabled: false,
    draggedId: null
});
/** @hidden */
var useGanttViewDependencyDragContext = function () { return React.useContext(exports.GanttViewDependencyDragContext); };
exports.useGanttViewDependencyDragContext = useGanttViewDependencyDragContext;
exports.GanttViewDependencyDragContext.displayName = 'KendoReactGanttViewDependencyDragContext';
/** @hidden */
var GanttViewContext = function (_a) {
    var children = _a.children, tasks = _a.tasksStore, timelineWidth = _a.timelineWidth, dependencyDrag = _a.dependencyDrag;
    return (React.createElement(exports.GanttViewTasksContext.Provider, { value: tasks },
        React.createElement(exports.GanttViewTimelineWidthContext.Provider, { value: timelineWidth },
            React.createElement(exports.GanttViewDependencyDragContext.Provider, { value: dependencyDrag }, children))));
};
exports.GanttViewContext = GanttViewContext;
exports.GanttViewContext.displayName = 'KendoReactGanttViewContext';
