// @progress/kendo-react-gantt/dist/npm/views/GanttWeekView.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultProps = exports.GanttWeekView = void 0;
var React = require("react");
var PropTypes = require("prop-types");
var kendo_date_math_1 = require("@progress/kendo-date-math");
var BaseView_1 = require("../components/BaseView");
var utils_1 = require("../utils");
var GanttContext_1 = require("../context/GanttContext");
var constants_1 = require("../constants");
var messages_1 = require("../messages");
var kendo_react_intl_1 = require("@progress/kendo-react-intl");
/**
 * Represents the KendoReact WeekView Component.
 */
exports.GanttWeekView = React.forwardRef(function (props, ref) {
    var _a = props.slotWidth, slotWidth = _a === void 0 ? exports.defaultProps.slotWidth : _a, firstSlotRangeRef = props.firstSlotRangeRef, timelineHeaderCell = props.timelineHeaderCell;
    var timezone = (0, GanttContext_1.useGanttPropsContext)().timezone;
    var intl = (0, kendo_react_intl_1.useInternationalization)();
    var weekViewRef = React.useRef(null);
    var baseViewRef = React.useRef(null);
    var scrollIntoView = React.useCallback(function (options) {
        if (baseViewRef.current) {
            baseViewRef.current.scrollIntoView(options);
        }
    }, []);
    React.useImperativeHandle(weekViewRef, function () { return ({ scrollIntoView: scrollIntoView }); });
    React.useImperativeHandle(ref, function () { return weekViewRef.current; });
    var range = (0, GanttContext_1.useGanttDateRangeContext)();
    var workWeekStart = props.workWeekStart || exports.defaultProps.workWeekStart;
    var workWeekEnd = props.workWeekEnd || exports.defaultProps.workWeekEnd;
    var slotLevels = React.useMemo(function () {
        var workDays = (0, utils_1.getWorkDays)({
            workWeekStart: workWeekStart,
            workWeekEnd: workWeekEnd
        });
        var daySlots = [];
        var weekSlots = (0, utils_1.getWeekSlots)(range, { timezone: timezone }, intl);
        weekSlots.forEach(function (weekSlot) {
            var days = (0, utils_1.getDaySlots)(weekSlot.range, { workDays: workDays, timezone: timezone }, intl);
            daySlots.push.apply(daySlots, days);
        });
        if (firstSlotRangeRef && daySlots[0]) {
            firstSlotRangeRef.current = daySlots[0].range;
        }
        return [weekSlots, daySlots];
    }, [timezone, range.start.getTime(), range.end.getTime(), intl, workWeekStart, workWeekEnd]);
    return (React.createElement(BaseView_1.BaseView, { ref: baseViewRef, slotWidth: slotWidth, slotLevels: slotLevels, timelineHeaderCell: timelineHeaderCell }));
});
var weekViewDateRange = function (_a) {
    var intl = _a.intl, tasksStart = _a.tasksStart, tasksEnd = _a.tasksEnd, timezone = _a.timezone;
    var zonedTasksStart = kendo_date_math_1.ZonedDate.fromLocalDate(tasksStart, timezone);
    var zonedTasksEnd = kendo_date_math_1.ZonedDate.fromLocalDate(tasksEnd, timezone);
    var firstDate = (0, utils_1.getStartDate)((0, kendo_date_math_1.firstDayInWeek)((0, kendo_date_math_1.getDate)(zonedTasksStart), intl.firstDay()));
    var lastDate = (0, utils_1.getEndDate)((0, kendo_date_math_1.firstDayInWeek)((0, kendo_date_math_1.getDate)(zonedTasksEnd), intl.firstDay()), constants_1.DAYS_IN_WEEK_COUNT);
    var zonedStart = kendo_date_math_1.ZonedDate.fromUTCDate((0, utils_1.toUTCDateTime)(firstDate), timezone);
    var zonedEnd = kendo_date_math_1.ZonedDate.fromUTCDate((0, utils_1.toUTCDateTime)(lastDate), timezone);
    return {
        start: new Date(zonedStart.getTime()),
        end: new Date(zonedEnd.getTime()),
        zonedStart: zonedStart,
        zonedEnd: zonedEnd
    };
};
/** @hidden */
exports.defaultProps = {
    name: 'week',
    dateRange: weekViewDateRange,
    title: function (localization) { return localization.toLanguageString(messages_1.weekViewTitle, messages_1.messages[messages_1.weekViewTitle]); },
    slotWidth: constants_1.DEFAULT_COLUMN_WIDTH,
    workWeekStart: 1,
    workWeekEnd: 5,
    workDayStart: '08:00',
    workDayEnd: '17:00'
};
exports.GanttWeekView.defaultProps = exports.defaultProps;
exports.GanttWeekView.propTypes = {
    name: PropTypes.string,
    dateRange: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
    title: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
    slotWidth: PropTypes.number,
    workWeekStart: PropTypes.number,
    workWeekEnd: PropTypes.number,
    workDayStart: PropTypes.string,
    workDayEnd: PropTypes.string
};
exports.GanttWeekView.displayName = 'KendoReactGanttWeekView';
