// @progress/kendo-react-gantt/dist/npm/views/GanttMonthView.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultProps = exports.GanttMonthView = void 0;
var React = require("react");
var PropTypes = require("prop-types");
var kendo_date_math_1 = require("@progress/kendo-date-math");
var BaseView_1 = require("../components/BaseView");
var utils_1 = require("../utils");
var GanttContext_1 = require("../context/GanttContext");
var constants_1 = require("../constants");
var messages_1 = require("../messages");
var kendo_react_intl_1 = require("@progress/kendo-react-intl");
/**
 * Represents the KendoReact MonthView Component.
 */
exports.GanttMonthView = React.forwardRef(function (props, ref) {
    var _a = props.slotWidth, slotWidth = _a === void 0 ? exports.defaultProps.slotWidth : _a, firstSlotRangeRef = props.firstSlotRangeRef, timelineHeaderCell = props.timelineHeaderCell;
    var timezone = (0, GanttContext_1.useGanttPropsContext)().timezone;
    var intl = (0, kendo_react_intl_1.useInternationalization)();
    var monthViewRef = React.useRef(null);
    var baseViewRef = React.useRef(null);
    var scrollIntoView = React.useCallback(function (options) {
        if (baseViewRef.current) {
            baseViewRef.current.scrollIntoView(options);
        }
    }, []);
    React.useImperativeHandle(monthViewRef, function () { return ({ scrollIntoView: scrollIntoView }); });
    React.useImperativeHandle(ref, function () { return monthViewRef.current; });
    var range = (0, GanttContext_1.useGanttDateRangeContext)();
    var slotLevels = React.useMemo(function () {
        var weekSlots = [];
        var monthSlots = (0, utils_1.getMonthSlots)(range, { timezone: timezone }, intl);
        monthSlots.forEach(function (monthSlot) {
            var weeks = (0, utils_1.getWeekSlots)(monthSlot.range, { timezone: timezone }, intl);
            weekSlots.push.apply(weekSlots, weeks);
        });
        if (firstSlotRangeRef && weekSlots[0]) {
            firstSlotRangeRef.current = weekSlots[0].range;
        }
        return [monthSlots, weekSlots];
    }, [timezone, range.start.getTime(), range.end.getTime(), intl]);
    return (React.createElement(BaseView_1.BaseView, { ref: baseViewRef, slotWidth: slotWidth, slotLevels: slotLevels, timelineHeaderCell: timelineHeaderCell }));
});
var monthViewDateRange = function (_a) {
    var tasksStart = _a.tasksStart, tasksEnd = _a.tasksEnd, timezone = _a.timezone;
    var zonedTasksStart = kendo_date_math_1.ZonedDate.fromLocalDate(tasksStart, timezone);
    var zonedTasksEnd = kendo_date_math_1.ZonedDate.fromLocalDate(tasksEnd, timezone);
    var firstDay = (0, kendo_date_math_1.firstDayOfMonth)((0, kendo_date_math_1.getDate)(zonedTasksStart));
    var lastDay = (0, kendo_date_math_1.addDays)((0, kendo_date_math_1.lastDayOfMonth)((0, kendo_date_math_1.getDate)(zonedTasksEnd)), 1);
    var zonedStart = kendo_date_math_1.ZonedDate.fromUTCDate((0, utils_1.toUTCDateTime)(firstDay), timezone);
    var zonedEnd = kendo_date_math_1.ZonedDate.fromUTCDate((0, utils_1.toUTCDateTime)(lastDay), timezone);
    return {
        start: new Date(zonedStart.getTime()),
        end: new Date(zonedEnd.getTime()),
        zonedStart: zonedStart,
        zonedEnd: zonedEnd
    };
};
/** @hidden */
exports.defaultProps = {
    name: 'month',
    dateRange: monthViewDateRange,
    title: function (localization) { return localization.toLanguageString(messages_1.monthViewTitle, messages_1.messages[messages_1.monthViewTitle]); },
    slotWidth: constants_1.DEFAULT_COLUMN_WIDTH,
    workWeekStart: 1,
    workWeekEnd: 5,
    workDayStart: '08:00',
    workDayEnd: '17:00'
};
exports.GanttMonthView.defaultProps = exports.defaultProps;
exports.GanttMonthView.propTypes = {
    name: PropTypes.string,
    dateRange: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
    title: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
    slotWidth: PropTypes.number,
    workWeekStart: PropTypes.number,
    workWeekEnd: PropTypes.number,
    workDayStart: PropTypes.string,
    workDayEnd: PropTypes.string
};
exports.GanttMonthView.displayName = 'GanttMonthView';
