// @progress/kendo-react-gantt/dist/npm/views/GanttDayView.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultProps = exports.GanttDayView = void 0;
var React = require("react");
var PropTypes = require("prop-types");
var kendo_date_math_1 = require("@progress/kendo-date-math");
var BaseView_1 = require("../components/BaseView");
var utils_1 = require("../utils");
var GanttContext_1 = require("../context/GanttContext");
var messages_1 = require("../messages");
var kendo_react_intl_1 = require("@progress/kendo-react-intl");
var constants_1 = require("../constants");
/**
 * Represents the KendoReact DayView Component.
 */
exports.GanttDayView = React.forwardRef(function (props, ref) {
    var _a = props.slotWidth, slotWidth = _a === void 0 ? exports.defaultProps.slotWidth : _a, _b = props.slotDuration, slotDuration = _b === void 0 ? exports.defaultProps.slotDuration : _b, firstSlotRangeRef = props.firstSlotRangeRef, timelineHeaderCell = props.timelineHeaderCell;
    var timezone = (0, GanttContext_1.useGanttPropsContext)().timezone;
    var intl = (0, kendo_react_intl_1.useInternationalization)();
    var dayViewRef = React.useRef(null);
    var baseViewRef = React.useRef(null);
    var scrollIntoView = React.useCallback(function (options) {
        if (baseViewRef.current) {
            baseViewRef.current.scrollIntoView(options);
        }
    }, []);
    React.useImperativeHandle(dayViewRef, function () { return ({ scrollIntoView: scrollIntoView }); });
    React.useImperativeHandle(ref, function () { return dayViewRef.current; });
    var range = (0, GanttContext_1.useGanttDateRangeContext)();
    var workWeekStart = props.workWeekStart || exports.defaultProps.workWeekStart;
    var workWeekEnd = props.workWeekEnd || exports.defaultProps.workWeekEnd;
    var workDayStart = props.workDayStart || exports.defaultProps.workDayStart;
    var workDayEnd = props.workDayEnd || exports.defaultProps.workDayEnd;
    var slotLevels = React.useMemo(function () {
        var workDays = (0, utils_1.getWorkDays)({
            workWeekStart: workWeekStart,
            workWeekEnd: workWeekEnd
        });
        var hourSlots = [];
        var daySlots = (0, utils_1.getDaySlots)(range, { workDays: workDays, timezone: timezone }, intl);
        daySlots.forEach(function (daySlot) {
            var hours = (0, utils_1.getHourSlots)(daySlot.range, {
                workDayStart: workDayStart,
                workDayEnd: workDayEnd,
                timezone: timezone,
                slotDuration: slotDuration
            }, intl);
            daySlot.span = hours.length;
            hourSlots.push.apply(hourSlots, hours);
        });
        if (firstSlotRangeRef && hourSlots[0]) {
            firstSlotRangeRef.current = hourSlots[0].range;
        }
        return [daySlots, hourSlots];
    }, [timezone, range.start.getTime(), range.end.getTime(), intl, workWeekStart, workWeekEnd, workDayStart, workDayEnd]);
    return (React.createElement(BaseView_1.BaseView, { ref: baseViewRef, slotWidth: slotWidth, slotLevels: slotLevels, timelineHeaderCell: timelineHeaderCell }));
});
var dayViewDateRange = function (_a) {
    // e.g.:
    // Gantt timezone: America/New_York (-4h)
    // PC timezone: Europe/Sofia (+3h)
    // tasksStart: Sun Jun 01 2014 03:00:00 GMT+0300 (Eastern European Summer Time)
    var tasksStart = _a.tasksStart, tasksEnd = _a.tasksEnd, timezone = _a.timezone;
    // Sat May 31 2014 20:00:00 GMT-0400 (EDT)
    var zonedTasksStart = kendo_date_math_1.ZonedDate.fromLocalDate(tasksStart, timezone);
    var zonedTasksEnd = kendo_date_math_1.ZonedDate.fromLocalDate(tasksEnd, timezone);
    // Sat May 31 2014 00:00:00 GMT+0300 (Eastern European Summer Time)
    var firstDate = (0, utils_1.getStartDate)(zonedTasksStart);
    var lastDate = (0, utils_1.getEndDate)(zonedTasksEnd);
    var zonedStart = kendo_date_math_1.ZonedDate.fromUTCDate((0, utils_1.toUTCDateTime)(firstDate), timezone);
    var zonedEnd = kendo_date_math_1.ZonedDate.fromUTCDate((0, utils_1.toUTCDateTime)(lastDate), timezone);
    return {
        // Sat May 31 2014 07:00:00 GMT+0300 (Eastern European Summer Time)
        start: new Date(zonedStart.getTime()),
        end: new Date(zonedEnd.getTime()),
        // Sat May 31 2014 00:00:00 GMT-0400 (EDT)
        zonedStart: zonedStart,
        zonedEnd: zonedEnd
    };
};
/** @hidden */
exports.defaultProps = {
    name: 'day',
    dateRange: dayViewDateRange,
    title: function (localization) { return localization.toLanguageString(messages_1.dayViewTitle, messages_1.messages[messages_1.dayViewTitle]); },
    slotWidth: constants_1.DEFAULT_COLUMN_WIDTH,
    slotDuration: 60,
    workWeekStart: 1,
    workWeekEnd: 5,
    workDayStart: '08:00',
    workDayEnd: '17:00'
};
exports.GanttDayView.defaultProps = exports.defaultProps;
exports.GanttDayView.propTypes = {
    name: PropTypes.string,
    dateRange: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
    title: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
    slotWidth: PropTypes.number,
    slotDuration: PropTypes.number,
    workWeekStart: PropTypes.number,
    workWeekEnd: PropTypes.number,
    workDayStart: PropTypes.string,
    workDayEnd: PropTypes.string
};
exports.GanttDayView.displayName = 'KendoReactGanttDayView';
