// @progress/kendo-react-pivotgrid/dist/npm/messages/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.messages = exports.configuratorButtonLabel = exports.configuratorMissingFields = exports.configuratorEditorSearchPlaceholder = exports.configuratorEditorSearchTitle = exports.configuratorMeasuresTitle = exports.configuratorRowsTitle = exports.configuratorColumnsTitle = exports.configuratorApply = exports.configuratorCancel = exports.configuratorTitle = exports.fieldMenuApply = exports.fieldMenuReset = exports.emptyCellAriaLabel = void 0;
/**
 * @hidden
 */
exports.emptyCellAriaLabel = 'pivotgrid.emptyCellAriaLabel';
/**
 * @hidden
 */
exports.fieldMenuReset = 'pivotgrid.fieldMenuReset';
/**
 * @hidden
 */
exports.fieldMenuApply = 'pivotgrid.fieldMenuApply';
/**
 * @hidden
 */
exports.configuratorTitle = 'pivotgrid.configuratorTitle';
/**
 * @hidden
 */
exports.configuratorCancel = 'pivotgrid.configuratorCancel';
/**
 * @hidden
 */
exports.configuratorApply = 'pivotgrid.configuratorApply';
/**
 * @hidden
 */
exports.configuratorColumnsTitle = 'pivotgrid.configuratorColumnsTitle';
/**
 * @hidden
 */
exports.configuratorRowsTitle = 'pivotgrid.configuratorRowsTitle';
/**
 * @hidden
 */
exports.configuratorMeasuresTitle = 'pivotgrid.configuratorMeasuresTitle';
/**
 * @hidden
 */
exports.configuratorEditorSearchTitle = 'pivotgrid.configuratorEditorSearchTitle';
/**
 * @hidden
 */
exports.configuratorEditorSearchPlaceholder = 'pivotgrid.configuratorEditorSearchPlaceholder';
/**
 * @hidden
 */
exports.configuratorMissingFields = 'pivotgrid.configuratorMissingFields';
/**
 * @hidden
 */
exports.configuratorButtonLabel = 'pivotgrid.configuratorButtonLabel';
/**
 * @hidden
 */
exports.messages = (_a = {},
    _a[exports.emptyCellAriaLabel] = 'PivotGrid component. Use the arrow keys to navigate.',
    _a[exports.fieldMenuReset] = 'Reset',
    _a[exports.fieldMenuApply] = 'Apply',
    _a[exports.configuratorCancel] = 'Cancel',
    _a[exports.configuratorApply] = 'Apply',
    _a[exports.configuratorTitle] = 'Settings',
    _a[exports.configuratorEditorSearchTitle] = 'Fields',
    _a[exports.configuratorEditorSearchPlaceholder] = 'Search...',
    _a[exports.configuratorColumnsTitle] = 'Columns',
    _a[exports.configuratorRowsTitle] = 'Rows',
    _a[exports.configuratorMeasuresTitle] = 'Values',
    _a[exports.configuratorButtonLabel] = 'Change settings',
    _a[exports.configuratorMissingFields] = 'Select some fields to begin setup',
    _a);
