// @progress/kendo-react-upload/dist/es/UploadListSingleItem.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import * as React from 'react';
import { classNames, IconWrap } from '@progress/kendo-react-common';
import { ProgressBar } from '@progress/kendo-react-progressbars';
import { provideLocalizationService, registerForLocalization } from '@progress/kendo-react-intl';
import { messages, statusUploaded, statusUploadFailed } from './messages';
import utils from './utils/utils';
import { UploadListActionButton } from './UploadListActionButton';
import { fileAudioIcon, fileConfigIcon, fileDataIcon, fileIcon, fileImageIcon, filePdfIcon, filePresentationIcon, fileProgrammingIcon, fileTxtIcon, fileVideoIcon, fileZipIcon } from '@progress/kendo-svg-icons';
/**
 * @hidden
 */
var UploadListSingleItem = /** @class */ (function (_super) {
    __extends(UploadListSingleItem, _super);
    function UploadListSingleItem() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @hidden
     */
    UploadListSingleItem.prototype.getFileExtension = function (file) {
        return file.extension ? file.extension.substring(1) : '';
    };
    /**
     * @hidden
     */
    UploadListSingleItem.prototype.getFileValidationMessage = function (file, isUploadFailed) {
        var localizationService = provideLocalizationService(this);
        var validationMessage = '';
        if (isUploadFailed) {
            validationMessage = localizationService.toLanguageString(statusUploadFailed, messages[statusUploadFailed]);
        }
        else if (file.validationErrors && file.validationErrors.length > 0) {
            var messageKey = "upload.".concat(file.validationErrors[0]);
            validationMessage = localizationService.toLanguageString(messageKey, messages[messageKey]);
        }
        return validationMessage;
    };
    /**
     * @hidden
     */
    UploadListSingleItem.prototype.getFileExtensionIcon = function (file) {
        switch (file.extension) {
            case '.png':
            case '.jpg':
            case '.jpeg':
            case '.tiff':
            case '.bmp':
            case '.gif':
                return { name: 'file-image', icon: fileImageIcon };
            case '.mp3':
            case '.mp4':
            case '.wav':
                return { name: 'file-audio', icon: fileAudioIcon };
            case '.mkv':
            case '.webm':
            case '.flv':
            case '.gifv':
            case '.avi':
            case '.wmv':
                return { name: 'file-video', icon: fileVideoIcon };
            case '.txt':
                return { name: 'file-txt', icon: fileTxtIcon };
            case '.pdf':
                return { name: 'file-pdf', icon: filePdfIcon };
            case '.ppt':
            case '.pptx':
                return { name: 'file-presentation', icon: filePresentationIcon };
            case '.csv':
            case '.xls':
            case '.xlsx':
                return { name: 'file-data', icon: fileDataIcon };
            case '.html':
            case '.css':
            case '.js':
            case '.ts':
                return { name: 'file-programming', icon: fileProgrammingIcon };
            case '.exe':
                return { name: 'file-config', icon: fileConfigIcon };
            case '.zip':
            case '.rar':
                return { name: 'file-zip', icon: fileZipIcon };
            default:
                return { name: 'file', icon: fileIcon };
        }
    };
    /**
     * @hidden
     */
    UploadListSingleItem.prototype.renderValidationError = function (file, isUploadFailed) {
        return (React.createElement("span", { className: 'k-file-info', key: '2' },
            React.createElement("span", { className: 'k-file-name', title: file.name }, file.name),
            React.createElement("span", { className: 'k-file-validation-message' }, this.getFileValidationMessage(file, isUploadFailed))));
    };
    /**
     * @hidden
     */
    UploadListSingleItem.prototype.renderFileDetails = function (file) {
        var localizationService = provideLocalizationService(this);
        var successMessage = localizationService.toLanguageString(statusUploaded, messages[statusUploaded]);
        return (React.createElement("span", { className: 'k-file-info', key: '2' },
            React.createElement("span", { className: 'k-file-name', title: file.name }, file.name),
            file.progress !== 100 ?
                React.createElement("span", { className: 'k-file-size' }, utils.getTotalFilesSizeMessage([file])) :
                React.createElement("span", { className: 'k-file-validation-message' }, successMessage)));
    };
    /**
     * @hidden
     */
    UploadListSingleItem.prototype.render = function () {
        var _a = this.props, files = _a.files, others = __rest(_a, ["files"]);
        var file = files[0];
        var _b = utils.getFileStatus([file]), isUploaded = _b[1], isUploadFailed = _b[2], isUploadValidationFailed = _b[3];
        var showProgressBar = !isUploadValidationFailed && !isUploaded && !isUploadFailed;
        return (React.createElement("div", { className: classNames('k-file-single') },
            showProgressBar && React.createElement(ProgressBar, { value: file.progress || 0 }),
            React.createElement("span", { className: 'k-file-icon-wrapper', key: '1' },
                React.createElement(IconWrap, __assign({ className: 'k-file-icon' }, this.getFileExtensionIcon(file))),
                React.createElement("span", { className: 'k-file-state' })),
            isUploadValidationFailed || isUploadFailed ?
                this.renderValidationError(file, isUploadFailed) :
                this.renderFileDetails(file),
            React.createElement(UploadListActionButton, __assign({ uid: file.uid, status: file.status, progress: file.progress }, others))));
    };
    return UploadListSingleItem;
}(React.Component));
export { UploadListSingleItem };
registerForLocalization(UploadListSingleItem);
