// @progress/kendo-react-sortable/dist/npm/utils/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.relativeContextElement = exports.closest = exports.isFocusable = exports.hasClasses = exports.toClassList = exports.findIndex = exports.find = void 0;
var kendo_react_common_1 = require("@progress/kendo-react-common");
/**
 * @hidden
 */
var find = function (data, comparer) {
    // required for IE11
    for (var i = 0; i < data.length; i++) {
        if (comparer(data[i])) {
            return data[i];
        }
    }
};
exports.find = find;
/**
 * @hidden
 */
var findIndex = function (data, comparer) {
    // required for IE11
    for (var i = 0; i < data.length; i++) {
        if (comparer(data[i])) {
            return i;
        }
    }
    return -1;
};
exports.findIndex = findIndex;
/**
 * @hidden
 */
var toClassList = function (classNames) { return String(classNames).trim().split(' '); };
exports.toClassList = toClassList;
/**
 * @hidden
 */
var hasClasses = function (element, classNames) {
    var namesList = toClassList(classNames);
    return Boolean(toClassList(element.className).find(function (className) { return namesList.indexOf(className) >= 0; }));
};
exports.hasClasses = hasClasses;
/**
 * @hidden
 */
var FOCUSABLE_REGEX = /^(?:a|input|select|option|textarea|button|object)$/i;
/**
 * @hidden
 */
var isFocusable = function (element) {
    if (element.tagName) {
        var tagName = element.tagName.toLowerCase();
        var tabIndex = element.getAttribute('tabIndex');
        var skipTab = tabIndex === '-1';
        var focusable = tabIndex !== null && !skipTab;
        if (FOCUSABLE_REGEX.test(tagName)) {
            focusable = !element.disabled && !skipTab;
        }
        return focusable;
    }
    return false;
};
exports.isFocusable = isFocusable;
/**
 * @hidden
 */
var closest = function (node, predicate) {
    while (node && !predicate(node)) {
        node = node.parentNode;
    }
    return node;
};
exports.closest = closest;
/**
 * @hidden
 */
var HAS_RELATIVE_STACKING_CONTEXT = (0, kendo_react_common_1.hasRelativeStackingContext)();
/**
 * @hidden
 */
var relativeContextElement = function (element) {
    if (!element || !HAS_RELATIVE_STACKING_CONTEXT) {
        return null;
    }
    var node = element.parentElement;
    while (node) {
        if (window.getComputedStyle(node).transform !== 'none') {
            return node;
        }
        node = node.parentElement;
    }
};
exports.relativeContextElement = relativeContextElement;
