// @progress/kendo-react-treelist/dist/npm/messages/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.messages = exports.noRecords = exports.filterBeforeOrEqualOperator = exports.filterBeforeOperator = exports.filterAfterOperator = exports.filterAfterOrEqualOperator = exports.filterBooleanAll = exports.filterIsFalse = exports.filterIsTrue = exports.filterLtOperator = exports.filterLteOperator = exports.filterGtOperator = exports.filterGteOperator = exports.filterEndsWithOperator = exports.filterNotContainsOperator = exports.filterContainsOperator = exports.filterStartsWithOperator = exports.filterIsNotEmptyOperator = exports.filterIsEmptyOperator = exports.filterIsNotNullOperator = exports.filterIsNullOperator = exports.filterNotEqOperator = exports.filterEqOperator = exports.filterClearButton = void 0;
/**
 * @hidden
 */
exports.filterClearButton = 'treelist.filterClearButton';
/**
 * @hidden
 */
exports.filterEqOperator = 'treelist.filterEqOperator';
/**
 * @hidden
 */
exports.filterNotEqOperator = 'treelist.filterNotEqOperator';
/**
 * @hidden
 */
exports.filterIsNullOperator = 'treelist.filterIsNullOperator';
/**
 * @hidden
 */
exports.filterIsNotNullOperator = 'treelist.filterIsNotNullOperator';
/**
 * @hidden
 */
exports.filterIsEmptyOperator = 'treelist.filterIsEmptyOperator';
/**
 * @hidden
 */
exports.filterIsNotEmptyOperator = 'treelist.filterIsNotEmptyOperator';
/**
 * @hidden
 */
exports.filterStartsWithOperator = 'treelist.filterStartsWithOperator';
/**
 * @hidden
 */
exports.filterContainsOperator = 'treelist.filterContainsOperator';
/**
 * @hidden
 */
exports.filterNotContainsOperator = 'treelist.filterNotContainsOperator';
/**
 * @hidden
 */
exports.filterEndsWithOperator = 'treelist.filterEndsWithOperator';
/**
 * @hidden
 */
exports.filterGteOperator = 'treelist.filterGteOperator';
/**
 * @hidden
 */
exports.filterGtOperator = 'treelist.filterGtOperator';
/**
 * @hidden
 */
exports.filterLteOperator = 'treelist.filterLteOperator';
/**
 * @hidden
 */
exports.filterLtOperator = 'treelist.filterLtOperator';
/**
 * @hidden
 */
exports.filterIsTrue = 'treelist.filterIsTrue';
/**
 * @hidden
 */
exports.filterIsFalse = 'treelist.filterIsFalse';
/**
 * @hidden
 */
exports.filterBooleanAll = 'treelist.filterBooleanAll';
/**
 * @hidden
 */
exports.filterAfterOrEqualOperator = 'treelist.filterAfterOrEqualOperator';
/**
 * @hidden
 */
exports.filterAfterOperator = 'treelist.filterAfterOperator';
/**
 * @hidden
 */
exports.filterBeforeOperator = 'treelist.filterBeforeOperator';
/**
 * @hidden
 */
exports.filterBeforeOrEqualOperator = 'treelist.filterBeforeOrEqualOperator';
/**
 * @hidden
 */
exports.noRecords = 'treelist.noRecords';
/**
 * @hidden
 */
exports.messages = (_a = {},
    _a[exports.filterClearButton] = 'Clear',
    _a[exports.filterEqOperator] = 'Is equal to',
    _a[exports.filterNotEqOperator] = 'Is not equal to',
    _a[exports.filterIsNullOperator] = 'Is null',
    _a[exports.filterIsNotNullOperator] = 'Is not null',
    _a[exports.filterIsEmptyOperator] = 'Is empty',
    _a[exports.filterIsNotEmptyOperator] = 'Is not empty',
    _a[exports.filterStartsWithOperator] = 'Starts with',
    _a[exports.filterContainsOperator] = 'Contains',
    _a[exports.filterNotContainsOperator] = 'Does not contain',
    _a[exports.filterEndsWithOperator] = 'Ends with',
    _a[exports.filterGteOperator] = 'Is greater than or equal to',
    _a[exports.filterGtOperator] = 'Is greater than',
    _a[exports.filterLteOperator] = 'Is less than or equal to',
    _a[exports.filterLtOperator] = 'Is less than',
    _a[exports.filterIsTrue] = 'Is true',
    _a[exports.filterIsFalse] = 'Is false',
    _a[exports.filterBooleanAll] = '(All)',
    _a[exports.filterAfterOrEqualOperator] = 'Is after or equal to',
    _a[exports.filterAfterOperator] = 'Is after',
    _a[exports.filterBeforeOperator] = 'Is before',
    _a[exports.filterBeforeOrEqualOperator] = 'Is before or equal to',
    _a[exports.noRecords] = 'No records available',
    _a);
