// @progress/kendo-react-gantt/dist/npm/messages/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.messages = exports.deleteDialogTitle = exports.deleteConfirmation = exports.editorDelete = exports.addBelow = exports.addAbove = exports.addChild = exports.addTask = exports.editorValidationEnd = exports.editorValidationStart = exports.editorValidationRequired = exports.editorTaskComplete = exports.editorTaskEnd = exports.editorTaskStart = exports.editorTaskTitle = exports.editorTitle = exports.editorCancel = exports.editorSave = exports.noRecords = exports.filterBeforeOrEqualOperator = exports.filterBeforeOperator = exports.filterAfterOperator = exports.filterAfterOrEqualOperator = exports.filterBooleanAll = exports.filterIsFalse = exports.filterIsTrue = exports.filterLtOperator = exports.filterLteOperator = exports.filterGtOperator = exports.filterGteOperator = exports.filterEndsWithOperator = exports.filterNotContainsOperator = exports.filterContainsOperator = exports.filterStartsWithOperator = exports.filterIsNotEmptyOperator = exports.filterIsEmptyOperator = exports.filterIsNotNullOperator = exports.filterIsNullOperator = exports.filterNotEqOperator = exports.filterEqOperator = exports.filterClearButton = exports.yearViewTitle = exports.monthViewTitle = exports.dayViewTitle = exports.weekViewTitle = void 0;
/** @hidden */
exports.weekViewTitle = 'gantt.weekViewTitle';
/** @hidden */
exports.dayViewTitle = 'gantt.dayViewTitle';
/** @hidden */
exports.monthViewTitle = 'gantt.monthViewTitle';
/** @hidden */
exports.yearViewTitle = 'gantt.yearViewTitle';
/** @hidden */
exports.filterClearButton = 'gantt.filterClearButton';
/** @hidden */
exports.filterEqOperator = 'gantt.filterEqOperator';
/** @hidden */
exports.filterNotEqOperator = 'gantt.filterNotEqOperator';
/** @hidden */
exports.filterIsNullOperator = 'gantt.filterIsNullOperator';
/** @hidden */
exports.filterIsNotNullOperator = 'gantt.filterIsNotNullOperator';
/** @hidden */
exports.filterIsEmptyOperator = 'gantt.filterIsEmptyOperator';
/** @hidden */
exports.filterIsNotEmptyOperator = 'gantt.filterIsNotEmptyOperator';
/** @hidden */
exports.filterStartsWithOperator = 'gantt.filterStartsWithOperator';
/** @hidden */
exports.filterContainsOperator = 'gantt.filterContainsOperator';
/** @hidden */
exports.filterNotContainsOperator = 'gantt.filterNotContainsOperator';
/** @hidden */
exports.filterEndsWithOperator = 'gantt.filterEndsWithOperator';
/** @hidden */
exports.filterGteOperator = 'gantt.filterGteOperator';
/** @hidden */
exports.filterGtOperator = 'gantt.filterGtOperator';
/** @hidden */
exports.filterLteOperator = 'gantt.filterLteOperator';
/** @hidden */
exports.filterLtOperator = 'gantt.filterLtOperator';
/** @hidden */
exports.filterIsTrue = 'gantt.filterIsTrue';
/** @hidden */
exports.filterIsFalse = 'gantt.filterIsFalse';
/** @hidden */
exports.filterBooleanAll = 'gantt.filterBooleanAll';
/** @hidden */
exports.filterAfterOrEqualOperator = 'gantt.filterAfterOrEqualOperator';
/** @hidden */
exports.filterAfterOperator = 'gantt.filterAfterOperator';
/** @hidden */
exports.filterBeforeOperator = 'gantt.filterBeforeOperator';
/** @hidden */
exports.filterBeforeOrEqualOperator = 'gantt.filterBeforeOrEqualOperator';
/** @hidden */
exports.noRecords = 'gantt.noRecords';
/** @hidden */
exports.editorSave = 'gantt.editSave';
/** @hidden */
exports.editorCancel = 'gantt.editCancel';
/** @hidden */
exports.editorTitle = 'gantt.editorTitle';
/** @hidden */
exports.editorTaskTitle = 'gantt.editorTaskTitle';
/** @hidden */
exports.editorTaskStart = 'gantt.editorTaskStart';
/** @hidden */
exports.editorTaskEnd = 'gantt.editorTaskEnd';
/** @hidden */
exports.editorTaskComplete = 'gantt.editorTaskComplete';
/** @hidden */
exports.editorValidationRequired = 'gantt.editorValidationRequired';
/** @hidden */
exports.editorValidationStart = 'gantt.editorValidationStart';
/** @hidden */
exports.editorValidationEnd = 'gantt.editorValidationEnd';
/** @hidden */
exports.addTask = 'gantt.addTask';
/** @hidden */
exports.addChild = 'gantt.addChild';
/** @hidden */
exports.addAbove = 'gantt.addAbove';
/** @hidden */
exports.addBelow = 'gantt.addBelow';
/** @hidden */
exports.editorDelete = 'gantt.editorDelete';
/** @hidden */
exports.deleteConfirmation = 'gantt.deleteConfirmation';
/** @hidden */
exports.deleteDialogTitle = 'gantt.deleteDialogTitle';
/**
 * @hidden
 */
exports.messages = (_a = {},
    _a[exports.weekViewTitle] = 'Week',
    _a[exports.dayViewTitle] = 'Day',
    _a[exports.monthViewTitle] = 'Month',
    _a[exports.yearViewTitle] = 'Year',
    _a[exports.filterClearButton] = 'Clear',
    _a[exports.filterEqOperator] = 'Is equal to',
    _a[exports.filterNotEqOperator] = 'Is not equal to',
    _a[exports.filterIsNullOperator] = 'Is null',
    _a[exports.filterIsNotNullOperator] = 'Is not null',
    _a[exports.filterIsEmptyOperator] = 'Is empty',
    _a[exports.filterIsNotEmptyOperator] = 'Is not empty',
    _a[exports.filterStartsWithOperator] = 'Starts with',
    _a[exports.filterContainsOperator] = 'Contains',
    _a[exports.filterNotContainsOperator] = 'Does not contain',
    _a[exports.filterEndsWithOperator] = 'Ends with',
    _a[exports.filterGteOperator] = 'Is greater than or equal to',
    _a[exports.filterGtOperator] = 'Is greater than',
    _a[exports.filterLteOperator] = 'Is less than or equal to',
    _a[exports.filterLtOperator] = 'Is less than',
    _a[exports.filterIsTrue] = 'Is true',
    _a[exports.filterIsFalse] = 'Is false',
    _a[exports.filterBooleanAll] = '(All)',
    _a[exports.filterAfterOrEqualOperator] = 'Is after or equal to',
    _a[exports.filterAfterOperator] = 'Is after',
    _a[exports.filterBeforeOperator] = 'Is before',
    _a[exports.filterBeforeOrEqualOperator] = 'Is before or equal to',
    _a[exports.noRecords] = 'No records available',
    _a[exports.editorSave] = 'Save',
    _a[exports.editorCancel] = 'Cancel',
    _a[exports.editorTitle] = 'Task',
    _a[exports.editorTaskTitle] = 'Title',
    _a[exports.editorTaskStart] = 'Start',
    _a[exports.editorTaskEnd] = 'End',
    _a[exports.editorTaskComplete] = 'Complete',
    _a[exports.editorValidationRequired] = 'Field is required.',
    _a[exports.editorValidationStart] = 'Start time must be be before End time.',
    _a[exports.editorValidationEnd] = 'End time must be after Start time.',
    _a[exports.addTask] = 'Add Task',
    _a[exports.addChild] = 'Add Child',
    _a[exports.addAbove] = 'Add Above',
    _a[exports.addBelow] = 'Add Below',
    _a[exports.editorDelete] = 'Delete',
    _a[exports.deleteConfirmation] = 'Are you sure you want to delete this event?',
    _a[exports.deleteDialogTitle] = 'Delete Event',
    _a);
