// @progress/kendo-pivotgrid-common/dist/npm/local-data-binding/schemaDimensions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/** @hidden */
exports.createFlatSchemaDimensions = function (dimensions, measures) {
    var result = Object.keys(dimensions).map(function (dim) { return ({
        caption: dim,
        defaultHierarchy: dim,
        description: '',
        name: dim,
        uniqueName: dim,
        hierarchyUniqueName: dim,
        measure: true,
        type: 1 // https://github.com/telerik/kendo-pivotgrid-common/blob/develop/src/models/responseDiscover.ts#L12-L14
    }); });
    if (measures.length) {
        result.push({
            caption: 'Measures',
            children: measures.map(function (m) { return ({
                aggregator: '1',
                caption: String(m.name),
                defaultFormat: '',
                description: '',
                displayFolder: '',
                groupName: String(m.name),
                name: String(m.name),
                uniqueName: String(m.name)
            }); }),
            type: 2,
            description: '',
            name: 'Measures',
            uniqueName: '[Measures]'
        });
    }
    return result;
};
