// @progress/kendo-react-pivotgrid/dist/es/messages/index.js
var _a;
/**
 * @hidden
 */
export var emptyCellAriaLabel = 'pivotgrid.emptyCellAriaLabel';
/**
 * @hidden
 */
export var fieldMenuReset = 'pivotgrid.fieldMenuReset';
/**
 * @hidden
 */
export var fieldMenuApply = 'pivotgrid.fieldMenuApply';
/**
 * @hidden
 */
export var configuratorTitle = 'pivotgrid.configuratorTitle';
/**
 * @hidden
 */
export var configuratorCancel = 'pivotgrid.configuratorCancel';
/**
 * @hidden
 */
export var configuratorApply = 'pivotgrid.configuratorApply';
/**
 * @hidden
 */
export var configuratorColumnsTitle = 'pivotgrid.configuratorColumnsTitle';
/**
 * @hidden
 */
export var configuratorRowsTitle = 'pivotgrid.configuratorRowsTitle';
/**
 * @hidden
 */
export var configuratorMeasuresTitle = 'pivotgrid.configuratorMeasuresTitle';
/**
 * @hidden
 */
export var configuratorEditorSearchTitle = 'pivotgrid.configuratorEditorSearchTitle';
/**
 * @hidden
 */
export var configuratorEditorSearchPlaceholder = 'pivotgrid.configuratorEditorSearchPlaceholder';
/**
 * @hidden
 */
export var configuratorMissingFields = 'pivotgrid.configuratorMissingFields';
/**
 * @hidden
 */
export var configuratorButtonLabel = 'pivotgrid.configuratorButtonLabel';
/**
 * @hidden
 */
export var messages = (_a = {},
    _a[emptyCellAriaLabel] = 'PivotGrid component. Use the arrow keys to navigate.',
    _a[fieldMenuReset] = 'Reset',
    _a[fieldMenuApply] = 'Apply',
    _a[configuratorCancel] = 'Cancel',
    _a[configuratorApply] = 'Apply',
    _a[configuratorTitle] = 'Settings',
    _a[configuratorEditorSearchTitle] = 'Fields',
    _a[configuratorEditorSearchPlaceholder] = 'Search...',
    _a[configuratorColumnsTitle] = 'Columns',
    _a[configuratorRowsTitle] = 'Rows',
    _a[configuratorMeasuresTitle] = 'Values',
    _a[configuratorButtonLabel] = 'Change settings',
    _a[configuratorMissingFields] = 'Select some fields to begin setup',
    _a);
