// @progress/kendo-react-gantt/dist/npm/constants/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_DEPENDENCY_MODEL_FIELDS = exports.DEFAULT_TASK_MODEL_FIELDS = exports.TASK_ID_ATT = exports.DEPENDENCY_DRAG_HANDLE_FINISH_ATT = exports.DEPENDENCY_DRAG_HANDLE_START_ATT = exports.DRAG_DIRECTION_FINISH = exports.DRAG_DIRECTION_START = exports.DEPENDENCY_DRAG_HANDLE = exports.MILESTONE_OFFSET = exports.DEFAULT_COLUMN_WIDTH = exports.YEAR_DATE_FORMAT = exports.MONTH_DATE_FORMAT = exports.DAY_DATE_FORMAT = exports.HOUR_DATE_FORMAT = exports.DAYS_IN_WEEK_COUNT = exports.DEFAULT_DATE = exports.MS_PER_DAY = exports.MS_PER_MINUTE = exports.MS_PER_HOUR = exports.GANTT_PREVENT_SELECTION_ELEMENT = exports.GANTT_COL_INDEX_ATTRIBUTE = exports.GANTT_ROW_INDEX_ATTRIBUTE = void 0;
var kendo_date_math_1 = require("@progress/kendo-date-math");
Object.defineProperty(exports, "MS_PER_DAY", { enumerable: true, get: function () { return kendo_date_math_1.MS_PER_DAY; } });
Object.defineProperty(exports, "MS_PER_HOUR", { enumerable: true, get: function () { return kendo_date_math_1.MS_PER_HOUR; } });
Object.defineProperty(exports, "MS_PER_MINUTE", { enumerable: true, get: function () { return kendo_date_math_1.MS_PER_MINUTE; } });
var kendo_react_data_tools_1 = require("@progress/kendo-react-data-tools");
/** The attribute required by the Gantt selection on Gantt `tr` elements. */
exports.GANTT_ROW_INDEX_ATTRIBUTE = kendo_react_data_tools_1.TABLE_ROW_INDEX_ATTRIBUTE;
/** The attribute required by the Gantt selection on Gantt `td` elements. */
exports.GANTT_COL_INDEX_ATTRIBUTE = kendo_react_data_tools_1.TABLE_COL_INDEX_ATTRIBUTE;
/** Attribute which disable selection start from this element. */
exports.GANTT_PREVENT_SELECTION_ELEMENT = kendo_react_data_tools_1.TABLE_PREVENT_SELECTION_ELEMENT;
/** @hidden */
exports.DEFAULT_DATE = new Date().getTime();
/** @hidden */
exports.DAYS_IN_WEEK_COUNT = 7;
/** @hidden */
exports.HOUR_DATE_FORMAT = 'hh:mm a';
/** @hidden */
exports.DAY_DATE_FORMAT = { skeleton: 'MEd' };
/** @hidden */
exports.MONTH_DATE_FORMAT = 'MMM';
/** @hidden */
exports.YEAR_DATE_FORMAT = 'yyyy';
/** @hidden */
exports.DEFAULT_COLUMN_WIDTH = 100;
/** @hidden */
exports.MILESTONE_OFFSET = 6;
/** @hidden */
exports.DEPENDENCY_DRAG_HANDLE = 'data-dependency-drag-handle';
/** @hidden */
exports.DRAG_DIRECTION_START = 'start';
/** @hidden */
exports.DRAG_DIRECTION_FINISH = 'finish';
/** @hidden */
exports.DEPENDENCY_DRAG_HANDLE_START_ATT = (_a = {}, _a[exports.DEPENDENCY_DRAG_HANDLE] = exports.DRAG_DIRECTION_START, _a);
/** @hidden */
exports.DEPENDENCY_DRAG_HANDLE_FINISH_ATT = (_b = {}, _b[exports.DEPENDENCY_DRAG_HANDLE] = exports.DRAG_DIRECTION_FINISH, _b);
/** @hidden */
exports.TASK_ID_ATT = 'data-task-id';
/**
 * @hidden
 */
exports.DEFAULT_TASK_MODEL_FIELDS = {
    id: 'id',
    start: 'start',
    end: 'end',
    title: 'title',
    percentComplete: 'percentComplete',
    isRollup: 'isRollup',
    isExpanded: 'isExpanded',
    isInEdit: 'isInEdit',
    children: 'children',
    isSelected: 'isSelected'
};
/**
 * @hidden
 */
exports.DEFAULT_DEPENDENCY_MODEL_FIELDS = {
    id: 'id',
    fromId: 'fromId',
    toId: 'toId',
    type: 'type'
};
