// @progress/kendo-pivotgrid-common/dist/npm/responses/discoverParser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var responseParser_1 = require("./responseParser");
/**
 * @hidden
 */
function parseCubes(response) {
    var xmlDoc = new DOMParser().parseFromString(response, "text/xml");
    var rows = Array.from(xmlDoc.querySelectorAll("DiscoverResponse > return > root > row"))
        .map(function (row) { return ({
        name: responseParser_1.getPropertyValue(row, "CUBE_NAME"),
        caption: responseParser_1.getPropertyValue(row, "CUBE_CAPTION"),
        description: responseParser_1.getPropertyValue(row, "DESCRIPTION"),
        type: responseParser_1.getPropertyValue(row, "CUBE_TYPE")
    }); });
    return rows;
}
exports.parseCubes = parseCubes;
/**
 * @hidden
 */
function parseCatalogs(response) {
    var xmlDoc = new DOMParser().parseFromString(response, "text/xml");
    var rows = Array.from(xmlDoc.querySelectorAll("DiscoverResponse > return > root > row"))
        .map(function (row) { return ({
        name: responseParser_1.getPropertyValue(row, "CATALOG_NAME"),
        description: responseParser_1.getPropertyValue(row, "DESCRIPTION")
    }); });
    return rows;
}
exports.parseCatalogs = parseCatalogs;
/**
 * @hidden
 */
function parseMeasures(response) {
    var xmlDoc = new DOMParser().parseFromString(response, "text/xml");
    var rows = Array.from(xmlDoc.querySelectorAll("DiscoverResponse > return > root > row"))
        .map(function (row) { return ({
        name: responseParser_1.getPropertyValue(row, "MEASURE_NAME"),
        caption: responseParser_1.getPropertyValue(row, "MEASURE_CAPTION"),
        uniqueName: responseParser_1.getPropertyValue(row, "MEASURE_UNIQUE_NAME"),
        description: responseParser_1.getPropertyValue(row, "DESCRIPTION"),
        aggregator: responseParser_1.getPropertyValue(row, "MEASURE_AGGREGATOR"),
        groupName: responseParser_1.getPropertyValue(row, "MEASUREGROUP_NAME"),
        displayFolder: responseParser_1.getPropertyValue(row, "MEASURE_DISPLAY_FOLDER"),
        defaultFormat: responseParser_1.getPropertyValue(row, "DEFAULT_FORMAT_STRING")
    }); });
    return rows;
}
exports.parseMeasures = parseMeasures;
/**
 * @hidden
 */
function parseKPIs(response) {
    var xmlDoc = new DOMParser().parseFromString(response, "text/xml");
    var rows = Array.from(xmlDoc.querySelectorAll("DiscoverResponse > return > root > row"))
        .map(function (row) { return ({
        name: responseParser_1.getPropertyValue(row, "KPI_NAME"),
        uniqueName: responseParser_1.getPropertyValue(row, "KPI_NAME"),
        caption: responseParser_1.getPropertyValue(row, "KPI_CAPTION"),
        value: responseParser_1.getPropertyValue(row, "KPI_VALUE"),
        goal: responseParser_1.getPropertyValue(row, "KPI_GOAL"),
        status: responseParser_1.getPropertyValue(row, "KPI_STATUS"),
        trend: responseParser_1.getPropertyValue(row, "KPI_TREND"),
        statusGraphic: responseParser_1.getPropertyValue(row, "KPI_STATUS_GRAPHIC"),
        trendGraphic: responseParser_1.getPropertyValue(row, "KPI_TREND_GRAPHIC"),
        description: responseParser_1.getPropertyValue(row, "KPI_DESCRIPTION"),
        groupName: responseParser_1.getPropertyValue(row, "MEASUREGROUP_NAME"),
        type: "kpi"
    }); });
    return rows;
}
exports.parseKPIs = parseKPIs;
/**
 * @hidden
 */
function parseDimensions(response) {
    var xmlDoc = new DOMParser().parseFromString(response, "text/xml");
    var rows = Array.from(xmlDoc.querySelectorAll("DiscoverResponse > return > root > row"))
        .map(function (row) { return ({
        caption: responseParser_1.getPropertyValue(row, "DIMENSION_CAPTION"),
        defaultHierarchy: responseParser_1.getPropertyValue(row, "DEFAULT_HIERARCHY"),
        description: responseParser_1.getPropertyValue(row, "DESCRIPTION"),
        name: responseParser_1.getPropertyValue(row, "DIMENSION_NAME"),
        type: parseInt(responseParser_1.getPropertyValue(row, "DIMENSION_TYPE"), 10),
        uniqueName: responseParser_1.getPropertyValue(row, "DIMENSION_UNIQUE_NAME")
    }); });
    return rows;
}
exports.parseDimensions = parseDimensions;
/**
 * @hidden
 */
function parseHierarchies(response) {
    var xmlDoc = new DOMParser().parseFromString(response, "text/xml");
    var rows = Array.from(xmlDoc.querySelectorAll("DiscoverResponse > return > root > row"))
        .map(function (row) { return ({
        name: responseParser_1.getPropertyValue(row, "HIERARCHY_NAME"),
        caption: responseParser_1.getPropertyValue(row, "HIERARCHY_CAPTION"),
        description: responseParser_1.getPropertyValue(row, "DESCRIPTION"),
        uniqueName: responseParser_1.getPropertyValue(row, "HIERARCHY_UNIQUE_NAME"),
        dimensionUniqueName: responseParser_1.getPropertyValue(row, "DIMENSION_UNIQUE_NAME"),
        displayFolder: responseParser_1.getPropertyValue(row, "HIERARCHY_DISPLAY_FOLDER"),
        origin: responseParser_1.getPropertyValue(row, "HIERARCHY_ORIGIN"),
        defaultMember: responseParser_1.getPropertyValue(row, "DEFAULT_MEMBER")
    }); });
    return rows;
}
exports.parseHierarchies = parseHierarchies;
/**
 * @hidden
 */
function parseLevels(response) {
    var xmlDoc = new DOMParser().parseFromString(response, "text/xml");
    var rows = Array.from(xmlDoc.querySelectorAll("DiscoverResponse > return > root > row"))
        .map(function (row) { return ({
        name: responseParser_1.getPropertyValue(row, "LEVEL_NAME"),
        caption: responseParser_1.getPropertyValue(row, "LEVEL_CAPTION"),
        description: responseParser_1.getPropertyValue(row, "DESCRIPTION"),
        uniqueName: responseParser_1.getPropertyValue(row, "LEVEL_UNIQUE_NAME"),
        dimensionUniqueName: responseParser_1.getPropertyValue(row, "DIMENSION_UNIQUE_NAME"),
        displayFolder: responseParser_1.getPropertyValue(row, "LEVEL_DISPLAY_FOLDER"),
        orderingProperty: responseParser_1.getPropertyValue(row, "LEVEL_ORDERING_PROPERTY"),
        origin: responseParser_1.getPropertyValue(row, "LEVEL_ORIGIN"),
        hierarchyUniqueName: responseParser_1.getPropertyValue(row, "HIERARCHY_UNIQUE_NAME")
    }); });
    return rows;
}
exports.parseLevels = parseLevels;
/**
 * @hidden
 */
function parseMembers(response) {
    var xmlDoc = new DOMParser().parseFromString(response, "text/xml");
    var rows = Array.from(xmlDoc.querySelectorAll("DiscoverResponse > return > root > row"))
        .map(function (row) { return ({
        name: responseParser_1.getPropertyValue(row, "MEMBER_NAME"),
        caption: responseParser_1.getPropertyValue(row, "MEMBER_CAPTION"),
        uniqueName: responseParser_1.getPropertyValue(row, "MEMBER_UNIQUE_NAME"),
        dimensionUniqueName: responseParser_1.getPropertyValue(row, "DIMENSION_UNIQUE_NAME"),
        hierarchyUniqueName: responseParser_1.getPropertyValue(row, "HIERARCHY_UNIQUE_NAME"),
        levelUniqueName: responseParser_1.getPropertyValue(row, "LEVEL_UNIQUE_NAME"),
        childrenCardinality: responseParser_1.getPropertyValue(row, "CHILDREN_CARDINALITY")
    }); });
    return rows;
}
exports.parseMembers = parseMembers;
