// @progress/kendo-react-indicators/dist/npm/skeleton/Skeleton.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Skeleton = void 0;
var React = require("react");
var PropTypes = require("prop-types");
var kendo_react_common_1 = require("@progress/kendo-react-common");
var kendo_react_common_2 = require("@progress/kendo-react-common");
var package_metadata_1 = require("../package-metadata");
/**
 * Represents the [KendoReact Skeleton component]({% slug overview_skeleton %}).
 *
 * @example
 * ```jsx
 * const App = () => {
 *   return (
 *       <Skeleton shape={'text'} style={{width: 200}}/>
 *   );
 * };
 *    ReactDOM.render(<App />, document.querySelector('my-app'));
 * ```
 */
exports.Skeleton = React.forwardRef(function (props, target) {
    (0, kendo_react_common_2.validatePackage)(package_metadata_1.packageMetadata);
    var className = props.className, style = props.style;
    var elementRef = React.useRef(null);
    var focusElement = React.useCallback(function () {
        if (elementRef.current) {
            elementRef.current.focus();
        }
    }, []);
    React.useImperativeHandle(target, function () { return ({
        element: elementRef.current,
        focus: focusElement
    }); });
    var shape = React.useMemo(function () { return props.shape || defaultProps.shape; }, [props.shape]);
    var animation = React.useMemo(function () {
        if (props.animation === false) {
            return false;
        }
        else {
            return props.animation || defaultProps.animation;
        }
    }, [props.animation]);
    var skeletonClasses = React.useMemo(function () { return (0, kendo_react_common_1.classNames)('k-skeleton', {
        'k-skeleton-circle': shape === 'circle',
        'k-skeleton-rectangle': shape === 'rectangle',
        'k-skeleton-text': shape === 'text',
        'k-skeleton-pulse': (typeof animation === 'object' && animation.type === 'pulse'),
        'k-skeleton-wave': (typeof animation === 'object' && animation.type === 'wave') || (animation === true)
    }, className); }, [shape, animation, className]);
    return (React.createElement("span", { ref: elementRef, className: skeletonClasses, style: style }));
});
exports.Skeleton.propTypes = {
    className: PropTypes.string,
    style: PropTypes.object,
    animation: PropTypes.shape({
        type: PropTypes.oneOf(['wave', 'pulse'])
    }),
    shape: PropTypes.oneOf(['circle', 'rectangle', 'text'])
};
var defaultProps = {
    shape: 'text',
    animation: { type: 'pulse' }
};
exports.Skeleton.defaultProps = defaultProps;
exports.Skeleton.displayName = 'KendoSkeleton';
